protocol/0083-RFPR-on_chain_referral_program.md

```diff
@@ -202,27 +219,53 @@ message UpdateReferralSet{
+
 If a referral set is currently designated as a team, a party is able to effectively "disband" a team by updating their referral set and setting their `is_team` value to `False`. Note a team should only be "disbanded" and removed from leaderboards at the end of the current epoch after rewards have been distributed.
 
 ### Applying a referral code
 
 To apply a referral code and become a referee, a party must fulfil the following criteria:
 
 - party must not currently be a **referrer**
+- party must not currently be a **referee** associated with a valid referral set (Note: for a set to be valid, the sets referrer must be meeting the staking requirement.)
 
 To become a referee, a referee must submit a signed `ApplyReferralCode` transaction with the following fields:
 
 - `referral_code`: the referral code they wish to apply
+- `alias`: Optional string defining an alias which should be associated with the key applying the referral code and should appear on leaderboards in the keys place.
 
 ```protobuf
 message ApplyReferralCode{
     id: "mYr3f3rra1c0d3"
+    alias: "myPuBlicA1ias"
 }
 ```
 
-If a party is not currently a referee, they must immediately be added to the referral set and [benefit factors and reward multipliers updated](#setting-benefit-factors-and-reward-multipliers) accordingly. Their key must then become associated with the referrer's key. All referral rewards will be transferred to this referrer's key, regardless of whether the party reapplies a new referral code.
+If a party is not currently a referee, they must immediately be added to the referral set and [benefit factors and reward multipliers updated](#setting-benefit-factors-and-reward-multipliers) accordingly. If a team exists for this referral set, they will also be added to that team.
+
 
-If a party is already a referee, and submits another `ApplyReferralCode` transaction, they will not be transferred to the new referral set but they will be added to the associated team at the start of the next epoch (providing a team exists). Note, if the referee has submitted multiple transactions in an epoch, the referee will be added to the new team specified in the latest valid transaction.
+If a party is already a referee, and submits another `ApplyReferralCode` transaction, the transaction will be rejected unless the referrer of the current referral set is not meeting the [staking requirement](#creating-a-referral-set). In this case, the party will be removed from the current referral set and added to the new referral set.
+
+
+### Joining a team
+
+If a party is already a referee in a valid set, to join or move between teams, a party can submit a `JoinTeam` transaction with the following fields:
+
+- `id`: the id of the team they wish to join
+- `alias`: Optional string defining an alias which should be associated with the key and should appear on leaderboards in the keys place.
+
+```protobuf
+message JoinTeam{
+    id: "mYr3f3rra1c0d3"
+    alias: "myPuBlicA1ias"
```

Comment:
Why use the alias to join a team ?

Reply:
So users have a way to change their `alias`.

Unless you would prefer a separate set/update alias transaction?

Reply:
Oh I understood the alias the other way around.

I am wondering if that kind of mechanics is not better as a separate domain, such as a "party / game profile" where we would be free to extend later on and keep other API free from such aspect. 

That would create less coupling.

Reply:
I agree it may be worth it if we planned to add more fields but think it would probably be overcomplicating it otherwise.

@MM0819 do you think it is likely there might be other metadata a party might want to add to their key in the future with respect to teams?

Reply:
Also @ValentinTrinque joining a team and then updating your `alias` (name on leaderboards) would then be two transactions which whilst not horrible I think we want to avoid?

Reply:
It's also worth considering to *not* necessarily associated a 1-to-1 relation between a form with and an endpoint. The front-end can expose a single form that calls 2 APIs.

That might not be the cleanest solution, but I think it's still worth considering about domain extensibility, flexibility and API evolution.

Basically, what happen if you want to change your alias while staying in the team ?
- Should you _re-join_ the team with a different alias ? 
- What kind of event should we emit for that, for the data-node ?

To me, they are 2 different needs:
1. Joining a team
2. Aliasing a party, network wide for team, referral set, games, whatever.

Reply:
Surely the issue with having alias on the team is that if I ever swap teams i have to keep re-inputting my alias each time?

We have the concept in the sketches of a user profile, and a user profile "card" ..... plus you can choose to compete when not in a team and would want an alias there too I suspect.

I dont know right now if we definitely will have more metadata associated to a user profile, but i can certainly imagine it ... discord ID, twitter handle, maybe awards / ranks / achievements etc.

To me a seperate user metadata thing seems more appropriate.

Reply:
Resolving as "alias"/"profiles" delayed to future release.